doc/decisions/operation_sequences.md

```diff
@@ -0,0 +1,92 @@
+# KDB operation sequences
+
+## Problem
+
+The current public API allows various kind of sequences of `kdbGet` and `kdbSet` calls.
+There is no real guideline of allowed and/or prohibited sequences, other than `kdbGet` for a certain parent key must be called before `kdbSet`:
+
+- initially after `kdbOpen`
+- after conflict errors in `kdbSet`
+
+While researching ways to do change tracking, we discovered that some sequences are currently legal, but might be problematic.
+This issue not only effects change tracking, but may cause problems for other use cases too.
+
+```c
+Key * keyA = ...;
+Key * keyB = ...;
+// assume that neither A below B nor B below A
+KeySet * a = ...;
+KeySet * b = ...;
+kdbGet(kdb, a, keyA); // (A)
+kdbGet(kdb, b, keyB); // (B)
+kdbSet(kdb, a, keyA); // (C)
+```
+
+- At point `(A)` the plugins internally store the KeySet for key `a`
+- At point `(B)` the plugins internally override the stored the KeySet with the KeySet for key `b`
+- At point `(C)` they will generate a wrong changeset for key `a`
+
+## Constraints
+
+- As far as possible, we must check for illegal sequences and raise an error when a constraint is violated
+
+## Assumptions
+
+1.
+2.
+3.
+
+## Considered Alternatives
+
+1. Document which sequences are allowed without raising errors in all illegal cases.
+   Least-effort approach, but at least offers some transparency for developers.
+   Developers utilizing Elektra can still ignore it, and may cause problems in certain setups without knowing.
+
+2. Enforce that the `parentKey` used in `kdbSet` is below the one used in the last `kdbGet`.
+   If this is not the case, `kdbSet` will abort and report an error in `parentKey`.
+   Developers might still wrongly mix the sequences, but they will get an error and have to fix it.
+
+   - Adding a `Key * lastGetParent` to `struct _KDB`
+   - Doing `keyCopy (handle->lastGetParent, parentKey, KEY_CP_NAME)` in `kdbGet`
+   - And checking `keyIsSameOrBelow (handle->lastGetParent, parentKey) == 1` in `kdbSet`
+
+3. Enforce that `ks` used in `kdbSet` is the same as in the last `kdbGet`
+   Similar to (2), but we check for the pointer of the `KeySet` instead of having a copy of the parent key, which saves some memory.
+
+   - Adding a `KeySet * lastGetKs` to `struct _KDB`
+   - Doing `handle->lastGetKs = ks` in `kdbGet`
+   - And checking `lastGetKs == ks` in `kdbSet`
+
+4. Change the API so that a single instance of KDB can only contain a single instance of a `KeySet`.
+   This solution completely eliminates problematic sequences.
+   ```c
+   // same as before but uses handle->data instead of the ks argument
+   int kdbGet (KDB * handle, Key * parentKey);
+   // same as before but uses handle->data instead of the ks argument
+   int kdbSet (KDB * handle, Key * parentKey);
+
+   KeySet * kdbData (KDB * handle) {
+       return handle->data;
+   }
+   ```
+
```

Comment:
IIRC there was also some idea of storing each previous parent key somewhere, so that we neither have to change the API nor have more illegal sequences as we currently have. Did I misunderstand?

Reply:
Yes, this is a possibility. But we need to check every plugin whether it implicitly assumes that `kdbSet` always follows `kdbGet` for the same parent key. If it does, we then need to add this 'workaround'.

Reply:
I've now made this alternative 5.